core/include/scipp/core/multi_index.h

```diff
@@ -310,29 +220,164 @@ template <scipp::index N> class MultiIndex {
+    m_coord[last_dim] = m_shape[last_dim];
+
+    for (scipp::index data = 0; data < N; ++data) {
+      // Only one dim contributes, all others have coord = 0.
+      m_bin[data].m_bin_index = m_coord[last_dim] * m_stride[last_dim][data];
+      load_bin_params(data);
+    }
+  }
+
+  scipp::index flat_index(const scipp::index i_data, scipp::index begin_index,
+                          const scipp::index end_index) {
+    scipp::index res = 0;
+    for (; begin_index < end_index; ++begin_index) {
+      res += m_coord[begin_index] * m_stride[begin_index][i_data];
+    }
+    return res;
+  }
+
+  [[nodiscard]] auto coord_it(const scipp::index dim = 0) noexcept {
+    return m_coord.begin() + dim;
+  }
+
+  [[nodiscard]] auto coord_end() noexcept { return m_coord.begin() + m_ndim; }
+
+  [[nodiscard]] auto shape_it(const scipp::index dim = 0) noexcept {
+    return std::next(m_shape.begin(), dim);
+  }
+
+  [[nodiscard]] auto shape_end() noexcept { return m_shape.begin() + m_ndim; }
+
+  /// Current flat index into the operands.
   std::array<scipp::index, N> m_data_index = {};
   // This does *not* 0-init the inner arrays!
-  std::array<std::array<scipp::index, NDIM_MAX>, N> m_stride = {};
+  /// Stride for each operand in each dimension.
+  std::array<std::array<scipp::index, N>, NDIM_MAX> m_stride = {};
+  /// Current index in iteration dimensions for both bin and inner dims.
   std::array<scipp::index, NDIM_MAX + 1> m_coord = {};
+  /// Shape of the iteration dimensions for both bin and inner dims.
   std::array<scipp::index, NDIM_MAX + 1> m_shape = {};
-  /// End-sentinel, essentially the volume of the iteration dimensions.
-  scipp::index m_end_sentinel{1};
-  /// Number of dimensions within bin, NDIM_MAX if no bins.
-  scipp::index m_ndim_nested{NDIM_MAX + 1};
-  /// Stride in bins along dim referred to by indices, e.g., 2D bins
+  /// Total number of dimensions.
+  scipp::index m_ndim{0};
+  /// Number of dense dimensions, i.e. same as m_ndim when not binned,
+  /// else number of dims in bins.
+  scipp::index m_inner_ndim{0};
+  /// Stride from one bin to the next.
+  scipp::index m_bin_stride = {};
+  /// Index of dim referred to by bin indices to distinguish, e.g., 2D bins
   /// slicing along first or second dim.
-  scipp::index m_nested_stride = {};
-  /// Index of dim referred to by indices to distinguish, e.g., 2D bins
-  /// slicing along first or second dim.
-  scipp::index m_nested_dim_index = {};
+  /// -1 if not binned.
+  scipp::index m_nested_dim_index{-1};
+  /// Parameters of the currently loaded bins.
   std::array<BinIterator, N> m_bin = {};
-  scipp::index m_ndim;
 };
+
 template <class... StridesArgs>
 MultiIndex(const Dimensions &, const StridesArgs &...)
     -> MultiIndex<sizeof...(StridesArgs)>;
 template <class... Params>
 MultiIndex(const ElementArrayViewParams &, const Params &...)
     -> MultiIndex<sizeof...(Params) + 1>;
 
+extern template SCIPP_CORE_EXPORT_TEMPLATE_DECLARATION
```

Comment:
This may explain why you need the odd macros above. Can you just remove all these lines here (all the `extern template` stuff)?

Reply:
Seems to work without the `extern` and explicit instantiations and with the basic `SCIPP_CORE_EXPORT` macro on my machine. Let's see what Windows thinks about it...

Reply:
We use this in multiple places. There is no need to put lines like these in the header, just the explicit instantiation in the cpp is enough.

Reply:
For exported classes? MSVC does not like it. It does not seem to export the constructors without an explicit instantiation of the *class* in the cpp file and accompanying export macro.

Reply:
It looks like you forgot some export macros where you instantiate (in the .cpp)?

Reply:
Here is an example where we do it for classes: https://github.com/scipp/scipp/blob/ae5db1fe1cdbe05f79dbcc5f7faa0648fdaa0da0/dataset/map_view.cpp#L317-L318

Reply:
Note that you need to add the macros also if you only instantiate a single function such as a constructur.